### PR TITLE
Migrate Package repositories page to bootstrap

### PR DESCRIPTION
**Pull request by mdeniz**
Instead of showing complex icons we have simplified those to only show the actual status of it (either enabled or disabled).

image

We made this page responsive for able small devices user's to work here.

image

The actions that can be performed per flag are now inside a popover (needs to click on the icon flag instead of just hovering). This popover shows in the title the actual status of the flag (and if this is a default value or not).

image

Now the spinner is shown in place (replacing the icon of the flag) until it finish.

repositories_flags_spinner

repositories_flags_spinner2

Co-authored-by: Björn Geuken <bgeuken@suse.de>
Co-authored-by: Saray Cabrera Padrón <scabrerapadron@suse.de>
Co-authored-by: Lukas Krause <lkrause@suse.de>

### CODE REVIEW

**Comment by coolo** [comment]
I find it very dangerous to allow to edit the flags independently. Each of these change_flag/create_flag calls writes a _meta into the backend. And the slowest wins, not the one last triggered by the user.

**Comment by coolo** [comment]
And I for one will miss the information if x86_64 is build disabled because of a package or a project flag.

**Comment by ChrisBr** [comment]
Here is the old one for comparing:

selection_044

It looks very nice on the first glance :heart:

**Comment by ChrisBr** [comment]
> I find it very dangerous to allow to edit the flags independently. Each of these change_flag/create_flag calls writes a _meta into the backend. And the slowest wins, not the one last triggered by the user.

> And I for one will miss the information if x86_64 is build disabled because of a package or a project flag.

@coolo I wasn't involved in the migration but it was like this before, wasn't it?

**Comment by ChrisBr** [comment]
It's awesome that it seems that the tests work but I'm also a little bit concerned about it :smile: :thinking: :see_no_evil:

**Comment by mdeniz** [comment]
> I find it very dangerous to allow to edit the flags independently. Each of these change_flag/create_flag calls writes a _meta into the backend. And the slowest wins, not the one last triggered by the user.

It was like that before.

> And I for one will miss the information if x86_64 is build disabled because of a package or a project flag.

Now in the title of the flag popover says if this value is the defaullt (aka inherited either from the project of the default for the app) or nothing if it is set by the user

**Comment by ChrisBr** [comment]
> Instead of showing complex icons we have simplified those to only show the actual status of it (either enabled or disabled).

That makes sense to me :+1:

File: src/api/app/views/webui2/shared/_flag_popover.html.haml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+- if flag.new_record?
+  %div
+    = link_to({ controller: 'repositories', action: 'create_flag', project: project, package: package, flag: flag.flag,
```
Comment (ChrisBr) [inline src/api/app/views/webui2/shared/_flag_popover.html.haml]: [nitpick] Can we not use URL helpers here?

File: src/api/app/views/webui2/shared/_flag_popover.html.haml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+- if flag.new_record?
```
Comment (ChrisBr) [inline src/api/app/views/webui2/shared/_flag_popover.html.haml]: Just to understand it: How can it be a new record?

**Comment by coolo** [comment]
> Now in the title of the flag popover says if this value is the defaullt (aka inherited either from the project of the default for the app) or nothing if it is set by the user

I had to experiment on 
https://obs-reviewlab.opensuse.org/mdeniz-bootstrapmigrate_repositories_page/repositories/openSUSE:Factory:Staging:A/test to understand what you're saying.

File: src/api/app/views/webui2/shared/_flag_popover.html.haml (hunk 2 above)
Comment (ChrisBr) [reply src/api/app/views/webui2/shared/_flag_popover.html.haml]: [nitpick] The only difference between the if / else are the URLs, right? Can we not just compute the URL in the if/else or somewhere else and then the partial just has the presentation with whatever link we computed?

File: src/api/app/views/webui2/shared/_repositories.html.haml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      Debuginfo Flag
+      %span.flagerror
+    = render partial: 'shared/repositories_flag_table', locals: { flags: debuginfo, project: project, package: package }
+  .col
+    %h5
+      Use for Build Flag
+      %span.flagerror
+    = render partial: 'shared/repositories_flag_table', locals: { flags: useforbuild, project: project, package: package }
+:javascript
+  $(document).ready(function() {
```
Comment (ChrisBr) [inline src/api/app/views/webui2/shared/_repositories.html.haml]: Please move this into the assets

File: src/api/app/views/webui2/shared/_repositories_flag_table.html.haml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+.table-responsive
+  %table.table.table-hover
+    %thead.thead-light
+      %tr
+        %th.col-6 Repository
+        %th.col-2.text-center All
+        - @architectures.each do |architecture|
```
Comment (ChrisBr) [inline src/api/app/views/webui2/shared/_repositories_flag_table.html.haml]: If you put project / package as local variables why not architecture as well?

**Comment by coolo** [comment]
> @coolo I wasn't involved in the migration but it was like this before, wasn't it?

That regression is older, yes. But it's pretty obvious in the last animated gif provided. As I can't test webui2 without login: how does it behave without login? Because the most annoying misfeature of the current page is that you only learn that you have no permissions to change flags, once you try.  As I can see a can_modify in the changes - is this fixed?

File: src/api/app/views/webui2/shared/_repositories_flag_table_column.html.haml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+:ruby
+  title = flag.effective_status == 'disable' ? 'Disabled' : 'Enabled'
+  title += ' (default)' if flag.new_record?
+  content = String.new(render(partial: 'shared/flag_popover', locals: { flag: flag, project: project, package: package, remote: !flag.has_children }))
```
Comment (ChrisBr) [inline src/api/app/views/webui2/shared/_repositories_flag_table_column.html.haml]: Can you please describe again what was the issue here?

File: src/api/app/views/webui2/shared/_repositories_flag_table.html.haml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+.table-responsive
+  %table.table.table-hover
+    %thead.thead-light
+      %tr
+        %th.col-6 Repository
```
[the remark this answers is not in the document]
Comment (ChrisBr) [reply src/api/app/views/webui2/shared/_repositories_flag_table.html.haml]: selection_045

I'm currently checking the pattern library.

We should switch to https://getbootstrap.com/docs/4.0/utilities/sizing/ instead.

[the remark this answers is not in the document]
Comment (ChrisBr) [reply src/api/app/views/webui2/shared/_repositories_flag_table.html.haml]: This is how pattern library currently looks in chrome :cry: :

selection_043

File: src/api/app/views/webui2/shared/_flag_popover.html.haml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    = link_to({ controller: 'repositories', action: 'create_flag', project: project, package: package, flag: flag.flag,
+              status: 'disable', repository: flag.repo, architecture: flag.arch }, method: :post, remote: remote,
+              class: 'popover_flag_action', data: { flag_id: flag.fullname }) do
+      %i.fas.fa-ban.text-danger
+      Disable
+  .pt-2
+    = link_to({ controller: 'repositories', action: 'create_flag', project: project, package: package, flag: flag.flag,
+              status: 'enable', repository: flag.repo, architecture: flag.arch }, method: :post, remote: remote,
+              class: 'popover_flag_action', data: { flag_id: flag.fullname }) do
+      %i.fas.fa-check.text-primary
```
[the remark this answers is not in the document]
Comment (mdeniz) [reply src/api/app/views/webui2/shared/_flag_popover.html.haml]: Those enable and disable icons are for the action... not for the status

File: src/api/app/views/webui2/shared/_flag_popover.html.haml (hunk 2 above)
Comment (mdeniz) [reply src/api/app/views/webui2/shared/_flag_popover.html.haml]: > Just to understand it: How can it be a new record?

Not all the flags are in the DB, the majority is calculated base on the default value for the flag type or is based on the project one..., so for showing it we always create a new record flag if it is not in the DB.

Comment (mdeniz) [reply src/api/app/views/webui2/shared/_flag_popover.html.haml]: > [nitpick] The only difference between the if / else are the URLs, right? Can we not just compute the URL in the if/else or somewhere else and then the partial just has the presentation with whatever link we computed?

Not only the url really, also the icon and the text are different

Comment (ChrisBr) [reply src/api/app/views/webui2/shared/_flag_popover.html.haml]: Smart :+1:

File: src/api/app/views/webui2/shared/_repositories_flag_table.html.haml (hunk 4 above)
Comment (mdeniz) [reply src/api/app/views/webui2/shared/_repositories_flag_table.html.haml]: Sure, oversaw it :+1:

File: src/api/app/views/webui2/shared/_flag_popover.html.haml (hunk 2 above)
Comment (ChrisBr) [reply src/api/app/views/webui2/shared/_flag_popover.html.haml]: > Not only the url really, also the http method, the icon and the text are different
 
Now I see it, but still very similar. But fine for me leaving it like it is...

File: src/api/app/views/webui2/shared/_repositories_flag_table_column.html.haml (hunk 5 above)
Comment (saraycp) [reply src/api/app/views/webui2/shared/_repositories_flag_table_column.html.haml]: When we did

`content = render(partial: 'shared/flag_popover', locals: { flag: flag, project: project, package: package, remote: !flag.has_children })`

and passed `content` as data attribute for popover, the view was broken. We realized that `content` was not `String` but `ActiveSupport::SafeBuffer`. So the `data` value received something different from a `String`.

Neither Haml nor `ActiveSupport::SafeBuffer#to_s` convert it to String. We had two options to do so: `String.new(...)` or interpolation (which was an offense).

**Comment by mdeniz** [comment]
> It's awesome that it seems that the tests work but I'm also a little bit concerned about it smile thinking see_no_evil

That's maybe because we still use the same ids in the flags...

File: src/api/app/assets/javascripts/webui2/repositories.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+$(document).ready(function() {
+    $(document).on('click','.popover_flag_action', function() {
```
[the remark this answers is not in the document]
Comment (mdeniz) [reply src/api/app/assets/javascripts/webui2/repositories.js]: Yes, better, will do.

**Comment by ChrisBr** [comment]
> On Firefox, whenever clicking on one of the icons to enable/disable a repository, it "kicks" to the top of the view. To reproduce it, scroll a bit down, then click on an icon. You will see what I mean.

Same on chrome, forgot to report it :+1:

**Comment by mdeniz** [comment]
> On Firefox, whenever clicking on one of the icons to enable/disable a repository, it "kicks" to the top of the view. To reproduce it, scroll a bit down, then click on an icon. You will see what I mean.

Yes, will fix it.

> I think we should remove the duplication for enable and take default.

No, those are two different things, one is changing the value of the package flag and the other is removing the value of the package to take whatever is coming from the project...

File: src/api/app/views/webui2/shared/_repositories_flag_table_column.html.haml (hunk 5 above)
Comment (ChrisBr) [reply src/api/app/views/webui2/shared/_repositories_flag_table_column.html.haml]: Thanks for the explanation.

File: src/api/app/views/webui2/shared/_flag_popover.html.haml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        %i.fas.fa-ban.text-danger
+        Disable
+      - else
+        %i.fas.fa-check.text-primary
+        Enable
+  .pt-2
+    = link_to(remove_repository_flag_path(project: project, package: package, flag: flag),
+              method: :delete, remote: remote, class: 'popover_flag_action', data: { flag_id: flag.fullname }) do
+      %i.fas{ class: flag.default_status == 'disable' ? 'fa-ban text-danger' : 'fa-check text-primary' }
+      Take default (#{flag.default_status})
```
Comment (dmarcoux) [inline src/api/app/views/webui2/shared/_flag_popover.html.haml]: As discussed with @mdeniz, this is not super clear since it doesn't state where the default is coming from. In the case of a package, it comes from the project. When this view is rendered in the context of a project, it comes from the application.

This will be addressed in another PR. I created an issue for it #6019.

**Comment by krauselukas** [comment]
@ChrisBr yes, i used w-auto for the table, seems to work quite well 
screenshot-2018-10-4 repositories for apache2 - open build service
screenshot-2018-10-4 repositories for apache2 - open build service 1

File: src/api/app/views/webui2/shared/_repositories_flag_table_column.html.haml (hunk 5 above)
Comment (hennevogel) [reply src/api/app/views/webui2/shared/_repositories_flag_table_column.html.haml]: offenses can be disabled. which everybody from the future will understand. this, no one will understand.

**Comment by hennevogel** [comment]
>> And I for one will miss the information if x86_64 is build disabled because of a package or a project flag.

> Now in the title of the flag popover says if this value is the defaullt (aka inherited either from the project of the default for the app) or nothing if it is set by the user

Which still means you can't get an overview. You have to click through _every_ icon's popover and write down if it's the default or not.

**Comment by mdeniz** [comment]
> Which still means you can't get an overview. You have to click through every icon's popover and write down if it's the default or not.

Can we brainstorm somewhere else? to gather some idea of how to do this without adding visual complexity...

**Comment by mdeniz** [comment]
Ok, what do you think about this?

image

If you like it I will push the changes right away...

**Comment by ChrisBr** [comment]
> Ok, what do you think about this? Default values don't have the little dot.

LGTM

**Comment by krauselukas** [comment]
@mdeniz LGTM aswell, maybe there is way to align them with the icons without a dot.

**Comment by Conan-Kudo** [comment]
I actually would rather have something that indicates the inheritance of settings. As the workflow I use for managing the build service instance I have at my workplace depends heavily on it, it'd be problematic if I can't figure out at a glance how the setting was set.

**Comment by hellcp** [comment]
I would use backgrounds changing colour depending on state of inheritance actually, because both blue and green can indicate active state (where blue is less active, just info, so could indicate status following default), and both red and yellow can indicate inactive state (where yellow is less dangerous, just warning, so could indicate status following default).

Also, one thing I would recommend is using alternating backgrounds there, it's really hard to tell which line follows which repo.

**Comment by coolo** [comment]
OBS team gives out @Ana06's ruler for people who have problems following long lines :)
